File: keycloak-client/src/main/java/cc/coopersoft/accesstoken/keycloak/KeycloakTokenHolder.java
package cc.coopersoft.accesstoken.keycloak;

import android.util.Log;
import cc.coopersoft.accesstoken.api.*;
import cc.coopersoft.accesstoken.api.callback.AuthenticationCallback;
import cc.coopersoft.accesstoken.api.callback.CodeRequestCallback;
import cc.coopersoft.accesstoken.api.callback.VerificationCallback;
import cc.coopersoft.accesstoken.keycloak.model.KeycloakCodeResult;
import cc.coopersoft.accesstoken.keycloak.model.KeycloakErrorResult;
import cc.coopersoft.accesstoken.keycloak.model.KeycloakTokenInfo;
import com.google.gson.FieldNamingPolicy;
import com.google.gson.GsonBuilder;
import okhttp3.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.Date;
import java.util.Objects;
import java.util.Optional;

public class KeycloakTokenHolder implements PhoneSupportAccessTokenHolder {

  public static class Builder {
    private String authenticationCodeUrl = null;

    private String verificationCodeUrl = null;

    private String registrationCodeUrl = null;

    private String accessTokenUrl = null;



    private String host = "";

    private Integer port = null;

    private String realms = null;

    private String scheme = "https";

    private String clientId;

    private String clientSecret;

    private String scope = null;

    private final OkHttpClient.Builder okHttpClientBuilder;

    private KeycloakTokenStorage storage;


    public Builder() {
      okHttpClientBuilder = new OkHttpClient.Builder();
    }

    private String getRootUrl(){
      if (host == null || realms == null){
        throw new ExceptionInInitializerError("please set host info!");
      }

      return scheme + "://" +
              host + Optional.ofNullable(port).map(p -> ":" + port).orElse("") +
              "/auth/realms/" + realms + "/";
    }

    public String getAuthenticationCodeUrl(){
      return Optional.ofNullable(authenticationCodeUrl).orElse(getRootUrl() + "sms/authentication-code" );
    }

    public String getVerificationCodeUrl(){
      return Optional.ofNullable(verificationCodeUrl).orElse(getRootUrl() + "sms/verification-code");
    }

    public String getRegistrationCodeUrl(){
      return Optional.ofNullable(registrationCodeUrl).orElse(getRootUrl() + "sms/registration-code");
    }

    public String getAccessTokenUrl(){
      return Optional.ofNullable(accessTokenUrl).orElse(getRootUrl() + "protocol/openid-connect/token");
    }

    public KeycloakTokenStorage getStorage() {
      return storage;
    }


    public String getClientId() {
      return clientId;
    }

    public String getClientSecret() {
      return clientSecret;
    }

    public String getScope() {
      return scope;
    }

    public OkHttpClient.Builder okHttpClientBuilder() {
      return okHttpClientBuilder;
    }

    public Builder scheme(@NotNull String scheme){
      this.scheme = scheme;
      return this;
    }

    public Builder host(@Nullable String host){
      this.host = host;
      return this;
    }

    public Builder port(Integer port){
      this.port = port;
      return this;
    }

    public Builder realms(@Nullable String realms){
      this.realms = realms;
      return this;
    }

    public Builder authenticationCodeUrl(@Nullable String url){
      this.authenticationCodeUrl = url;
      return this;
    }

    public Builder verificationCodeUrl(@Nullable String url){
      this.verificationCodeUrl = url;
      return this;
    }

    public Builder registrationCodeUrl(@Nullable String url){
      this.registrationCodeUrl = url;
      return this;
    }

    public Builder accessTokenUrl(@Nullable String url){
      this.accessTokenUrl = url;
      return this;
    }

    public Builder clientId(@Nullable String clientId){
      this.clientId = clientId;
      return this;
    }

    public Builder clientSecret(String clientSecret){
      this.clientSecret = clientSecret;
      return this;
    }

    public Builder clientScope(String scope){
      this.scope = scope;
      return this;
    }

    public Builder storage(KeycloakTokenStorage storage){
      this.storage = storage;
      return this;
    }

    public KeycloakTokenHolder build(){
      return new KeycloakTokenHolder(this);
    }
  }

  private static final String LOG_TAG = "Keycloak";

  private KeycloakTokenInfo token;

  private OkHttpClient okHttpClient;
  private String authenticationCodeUrl;
  private String verificationCodeUrl;
  private String registrationCodeUrl;
  private String accessTokenUrl;
  private KeycloakTokenStorage storage;
  private String clientId;
  private String clientSecret;
  private String scope;


  KeycloakTokenHolder(Builder keycloakServer) {
    this.okHttpClient = keycloakServer.okHttpClientBuilder.build();
    this.authenticationCodeUrl = keycloakServer.getAuthenticationCodeUrl();
    this.verificationCodeUrl = keycloakServer.getVerificationCodeUrl();
    this.registrationCodeUrl = keycloakServer.getRegistrationCodeUrl();
    this.accessTokenUrl = keycloakServer.getAccessTokenUrl();
    this.clientId = keycloakServer.clientId;
    this.clientSecret = keycloakServer.clientSecret;
    this.scope = keycloakServer.scope;
    this.storage = keycloakServer.storage;
  }

  private KeycloakTokenHolder(){

  }

  private Optional<KeycloakTokenInfo> getToken(){
    if (token == null){
      token = storage.loadToken();
    }
    Log.d("Keycloak", "getToken:" + token);
    return Optional.ofNullable(token);
  }

  private boolean isExpires(long expiresIn, Date created){
    // sub one minute 1000 * 60
    return created.getTime() + expiresIn * 1000 - 1000 * 60 < System.currentTimeMillis();
  }

  private void saveToken(KeycloakTokenInfo token){
    this.token = token;
    storage.storeToken(token);
  }

//
//  private Optional<String> refreshToken(){
//    return getToken().flatMap(t -> isExpires(token.getRefreshExpiresIn(),token.getCreated()) ? Optional.empty() : Optional.of(t.getRefreshToken()));
//  }

  private Optional<KeycloakTokenInfo> validToken(@NotNull KeycloakTokenInfo token){
    if (isExpires(token.getExpiresIn(),token.getCreated())){
      Log.d("Keycloak", "token is expires" );
      if (isExpires(token.getRefreshExpiresIn(),token.getCreated())){
        Log.d("Keycloak", "refresh token is expires" );
        return Optional.empty();
      }
      return refreshToken(token.getRefreshToken());
    }
    return Optional.of(token);
  }

  @Override
  public Optional<String> getAccessToken(){
    return getToken()
            .flatMap(this::validToken)
            .map(t -> t.getTokenType() + " " + t.getAccessToken());
  }

  @Override
  public boolean hasToken(){
    return getToken().map(t -> {
      boolean expires = isExpires(token.getExpiresIn(),token.getCreated()) && isExpires(token.getRefreshExpiresIn(),token.getCreated());
      if (expires){
        clearToken();
      }
      return !expires;
    }).orElse(false);
  }

  @Override
  public void clearToken() {
    token = null;
    storage.removeToken();
  }

  @Override
  public void sendAuthenticationCode(String phoneNumber, CodeRequestCallback callback){
    sendCode(authenticationCodeUrl,phoneNumber, callback);
  }

  @Override
  public void sendRegistrationCode(String phoneNumber, CodeRequestCallback callback){
    sendCode(registrationCodeUrl,phoneNumber, callback);
  }

  @Override
  public void sendVerificationCode(String phoneNumber, CodeRequestCallback callback){
    sendCode(verificationCodeUrl,phoneNumber, callback);
  }

  @Override
  public void verificationCode(String phoneNumber, String code, VerificationCallback callback){

    Optional<String> accessToken = getAccessToken();
    if (!accessToken.isPresent()){
      callback.onError(phoneNumber, KeycloakErrorResult.builder().error("no token").errorDescription("must be have a token").build());
      return;
    }

    HttpUrl url = Objects.requireNonNull(HttpUrl.parse(verificationCodeUrl))
            .newBuilder()
            .addQueryParameter("phoneNumber",phoneNumber)
            .addQueryParameter("code",code)
            .build();
    RequestBody body = RequestBody.create(null, new byte[0]);
    Request request = new Request.Builder()
            .addHeader("authorization", getAccessToken().orElseThrow(IllegalArgumentException::new))
            .url(url)
            .post(body)
            .build();
    okHttpClient.newCall(request).enqueue(new Callback() {
      @Override
      public void onFailure(Call call, IOException e) {
        callback.onFailure(phoneNumber);
      }

      @Override
      public void onResponse(Call call, Response response) throws IOException {
        if (response.isSuccessful()){
          callback.onSuccess(phoneNumber);
        }else{
          callback.onFailure(phoneNumber);
        }
      }
    });

  }

  @Override
  public void requireToken(Credential credential, AuthenticationCallback callback){
    if (clientId == null){
      throw new ExceptionInInitializerError("client id not set!");
    }

    FormBody.Builder formBody = new FormBody.Builder();
    formBody.add("client_id",clientId)
            .add("grant_type","password");
    credential.getContext().forEach((k,v) -> formBody.add(k,v));
    Optional.ofNullable(clientSecret).ifPresent(s -> formBody.add("client_secret",s));
    Optional.ofNullable(scope).ifPresent(s -> formBody.add("scope",scope));
    Request request = new Request.Builder()
            .addHeader("Content-Type","application/x-www-form-urlencoded")
            .url(accessTokenUrl)
            .post(formBody.build())
            .build();
    okHttpClient.newCall(request).enqueue(new Callback() {
      @Override
      public void onFailure(Call call, IOException e) {
        Log.e(LOG_TAG, "failure" ,e);
        callback.onFailure(KeycloakErrorResult.builder().
                        error(e.getClass().getSimpleName()).
                        errorDescription(e.getMessage()).build());
      }

      @Override
      public void onResponse(Call call, Response response) throws IOException {
        if (response.isSuccessful()){
          assert response.body() != null;
          saveToken(parseToken(response.body().string()));
          callback.onSuccess();
        }else{
          if (response.body() != null){
            callback.onFailure(parseErrorResult(response.body().string()));
          }else{
            callback.onFailure(new KeycloakErrorResult("unknow","unknow error"));
          }
        }
      }
    });
  }

  @Override
  public Optional<String> refreshToken(){
    return getToken().flatMap(t -> refreshToken(token.getRefreshToken()))
            .map(t -> t.getTokenType() + " " + t.getAccessToken());
  }

  private Optional<KeycloakTokenInfo> refreshToken(String refreshToken){
    Log.d("Keycloak", "refresh token");
    if (clientId == null){
      throw new ExceptionInInitializerError("client id not set!");
    }
    Log.d("Keycloak","refresh token for:" + accessTokenUrl);
    FormBody.Builder formBody = new FormBody.Builder();
    formBody.add("client_id",clientId)
            .add("grant_type","refresh_token")
            .add("refresh_token", refreshToken);
    Optional.ofNullable(clientSecret).ifPresent(s -> formBody.add("client_secret",s));
    Optional.ofNullable(scope).ifPresent(s -> formBody.add("scope",scope));
    Request request = new Request.Builder()
            .addHeader("Content-Type","application/x-www-form-urlencoded")
            .url(accessTokenUrl)
            .post(formBody.build())
            .build();
    try {
      Response response = okHttpClient.newCall(request).execute();
      if (response.isSuccessful()) {
        Log.d(LOG_TAG, "token is refresh");
        assert response.body() != null;
        KeycloakTokenInfo token = parseToken(response.body().string());
        saveToken(token);
        return Optional.of(token);
      } else {
        if (response.body() != null) {
          Log.e(LOG_TAG,"refresh token fail! body is:" + response.body().string());
        } else {
          Log.e(LOG_TAG,"refresh token fail! body is null!");
        }
        //remove this token? maybe only network fail!
      }
    }catch (IOException e){
      Log.e(LOG_TAG,"refresh token fail! body is:" + e.getMessage(),e);
    }
    return Optional.empty();
  }


  private void sendCode(String requestUrl, String phoneNumber, CodeRequestCallback callback){
    HttpUrl url = Objects.requireNonNull(HttpUrl.parse(requestUrl)).newBuilder().addQueryParameter("phoneNumber",phoneNumber).build();
    Request request = new Request.Builder()
            .addHeader("Accept", "application/json; charset=utf-8")
            .addHeader("Cache-Control","no-cache")
            .url(url)
            .build();
    okHttpClient.newCall(request).enqueue(new Callback() {
      @Override
      public void onFailure(Call call, IOException e) {
        callback.onFailure(phoneNumber, new KeycloakErrorResult("unknow"));
      }

      @Override
      public void onResponse(Call call, Response response) throws IOException {
        if (response.isSuccessful()){
          assert response.body() != null;
          callback.onSuccess(phoneNumber,parseCodeResult(response.body().string()));
        }else{
          if (response.body() != null) {
            callback.onFailure(phoneNumber, parseErrorResult(response.body().string()));
          }else{
            callback.onFailure(phoneNumber, new KeycloakErrorResult("unknow"));
          }
        }
      }
    });
  }

  private KeycloakErrorResult parseErrorResult(String json){
    if (json == null || json.trim().length() <= 0){
      return new KeycloakErrorResult("unknow");
    }
    final GsonBuilder gsonBuilder = new GsonBuilder();
    gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES);
    gsonBuilder.setLenient();
    return gsonBuilder.create().fromJson(json, KeycloakErrorResult.class);
  }

  private KeycloakTokenInfo parseToken(String json){
    Log.d("Keycloak","parseToken" + json);
    final GsonBuilder gsonBuilder = new GsonBuilder();
    gsonBuilder.setLenient();
    gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES);
    KeycloakTokenInfo result = gsonBuilder.create().fromJson(json, KeycloakTokenInfo.class);
    result.setCreated(new Date());
    return result;
  }

  private KeycloakCodeResult parseCodeResult(String json){
    final GsonBuilder gsonBuilder = new GsonBuilder();
    gsonBuilder.setLenient();
    gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES);
    KeycloakCodeResult result = gsonBuilder.create().fromJson(json, KeycloakCodeResult.class);
    result.setCreated(new Date());
    return result;
  }

}

File: keycloak-client/src/main/java/cc/coopersoft/accesstoken/keycloak/KeycloakTokenHolder.kt
package cc.coopersoft.accesstoken.keycloak

import android.util.Log
import cc.coopersoft.accesstoken.api.Credential
import cc.coopersoft.accesstoken.api.PhoneSupportAccessTokenHolder
import cc.coopersoft.accesstoken.api.callback.AuthenticationCallback
import cc.coopersoft.accesstoken.api.callback.CodeRequestCallback
import cc.coopersoft.accesstoken.api.callback.VerificationCallback
import cc.coopersoft.accesstoken.keycloak.model.KeycloakCodeResult
import cc.coopersoft.accesstoken.keycloak.model.KeycloakErrorResult
import cc.coopersoft.accesstoken.keycloak.model.KeycloakTokenInfo
import com.google.gson.FieldNamingPolicy
import com.google.gson.GsonBuilder
import okhttp3.Call
import okhttp3.Callback
import okhttp3.FormBody
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.Response
import java.io.IOException
import java.lang.NullPointerException
import java.util.Date
import java.util.Optional
import java.util.function.BiConsumer

class KeycloakTokenHolder : PhoneSupportAccessTokenHolder {
    class Builder {
        private var authenticationCodeUrl: String? = null
        private var verificationCodeUrl: String? = null
        private var registrationCodeUrl: String? = null
        private var accessTokenUrl: String? = null
        //relative path after domain, before realms
        private var preRealmsPath: String = "/auth"
        private var host: String? = ""
        private var port: Int? = null
        private var realms: String? = null
        private var scheme = "https"
        var clientId: String? = null
            private set
        var clientSecret: String? = null
            private set
        var scope: String? = null
            private set
        val okHttpClientBuilder: OkHttpClient.Builder = OkHttpClient.Builder()
        var storage: KeycloakTokenStorage? = null
            private set

        private val rootUrl: String
            get() {
                if (host == null || realms == null) {
                    throw ExceptionInInitializerError("please set host info!")
                }
                return scheme + "://" +
                        host + Optional.ofNullable(port).map { p: Int? -> ":$port" }
                    .orElse("") +
                        "$preRealmsPath/realms/" + realms + "/"
            }

        fun getAuthenticationCodeUrl(): String {
            return Optional.ofNullable(authenticationCodeUrl)
                .orElse(rootUrl + "sms/authentication-code")
        }

        fun getVerificationCodeUrl(): String {
            return Optional.ofNullable(verificationCodeUrl)
                .orElse(rootUrl + "sms/verification-code")
        }

        fun getRegistrationCodeUrl(): String {
            return Optional.ofNullable(registrationCodeUrl)
                .orElse(rootUrl + "sms/registration-code")
        }

        fun getAccessTokenUrl(): String {
            return Optional.ofNullable(accessTokenUrl)
                .orElse(rootUrl + "protocol/openid-connect/token")
        }

        fun okHttpClientBuilder(): OkHttpClient.Builder {
            return okHttpClientBuilder
        }

        fun scheme(scheme: String): Builder {
            this.scheme = scheme
            return this
        }

        fun host(host: String?): Builder {
            this.host = host
            return this
        }

        fun port(port: Int?): Builder {
            this.port = port
            return this
        }

        /**
         * Relative path after domain, before realms
         *
         * @param preRealmsPath Relative path after domain, null means no relative path
         */
        fun preRealmsPath(preRealmsPath: String?): Builder {
            this.preRealmsPath = preRealmsPath ?: ""
            return this
        }

        fun realms(realms: String?): Builder {
            this.realms = realms
            return this
        }

        fun authenticationCodeUrl(url: String?): Builder {
            authenticationCodeUrl = url
            return this
        }

        fun verificationCodeUrl(url: String?): Builder {
            verificationCodeUrl = url
            return this
        }

        fun registrationCodeUrl(url: String?): Builder {
            registrationCodeUrl = url
            return this
        }

        fun accessTokenUrl(url: String?): Builder {
            accessTokenUrl = url
            return this
        }

        fun clientId(clientId: String?): Builder {
            this.clientId = clientId
            return this
        }

        fun clientSecret(clientSecret: String?): Builder {
            this.clientSecret = clientSecret
            return this
        }

        fun clientScope(scope: String?): Builder {
            this.scope = scope
            return this
        }

        fun storage(storage: KeycloakTokenStorage?): Builder {
            this.storage = storage
            return this
        }

        fun build(): KeycloakTokenHolder {
            return KeycloakTokenHolder(this)
        }
    }

    private var token: KeycloakTokenInfo? = null
    private var okHttpClient: OkHttpClient? = null
    private var authenticationCodeUrl: String? = null
    private var verificationCodeUrl: String? = null
    private var registrationCodeUrl: String? = null
    private var accessTokenUrl: String? = null
    private var storage: KeycloakTokenStorage? = null
    private var clientId: String? = null
    private var clientSecret: String? = null
    private var scope: String? = null

    internal constructor(keycloakServer: Builder) {
        okHttpClient = keycloakServer.okHttpClientBuilder.build()
        authenticationCodeUrl = keycloakServer.getAuthenticationCodeUrl()
        verificationCodeUrl = keycloakServer.getVerificationCodeUrl()
        registrationCodeUrl = keycloakServer.getRegistrationCodeUrl()
        accessTokenUrl = keycloakServer.getAccessTokenUrl()
        clientId = keycloakServer.clientId
        clientSecret = keycloakServer.clientSecret
        scope = keycloakServer.scope
        storage = keycloakServer.storage
    }

    private constructor() {}

    private fun getToken(): Optional<KeycloakTokenInfo> {
        if (token == null) {
            token = storage!!.loadToken()
        }
        Log.d("Keycloak", "getToken:$token")
        return Optional.ofNullable(token)
    }

    private fun isExpires(expiresIn: Long, created: Date?): Boolean {
        // sub one minute 1000 * 60
        return created!!.time + expiresIn * 1000 - 1000 * 60 < System.currentTimeMillis()
    }

    private fun saveToken(token: KeycloakTokenInfo) {
        this.token = token
        storage!!.storeToken(token)
    }

    //
    //  private Optional<String> refreshToken(){
    //    return getToken().flatMap(t -> isExpires(token.getRefreshExpiresIn(),token.getCreated()) ? Optional.empty() : Optional.of(t.getRefreshToken()));
    //  }
    private fun validToken(token: KeycloakTokenInfo): Optional<KeycloakTokenInfo> {
        if (isExpires(token.expiresIn, token.created)) {
            Log.d("Keycloak", "token is expires")
            if (isExpires(token.refreshExpiresIn, token.created)) {
                Log.d("Keycloak", "refresh token is expires")
                return Optional.empty()
            }
            return refreshToken(token.refreshToken)
        }
        return Optional.of(token)
    }

    override val accessToken: Optional<String>
        get() = getToken()
            .flatMap { token: KeycloakTokenInfo -> validToken(token) }
            .map { t: KeycloakTokenInfo -> t.tokenType + " " + t.accessToken }

    override fun hasToken(): Boolean {
        return getToken().map { t: KeycloakTokenInfo ->
            val expires = isExpires(t.expiresIn, t.created) && isExpires(
                t.refreshExpiresIn, t.created
            )
            if (expires) {
                clearToken()
            }
            !expires
        }.orElse(false)
    }

    override fun clearToken() {
        token = null
        storage!!.removeToken()
    }

    override fun sendAuthenticationCode(phoneNumber: String?, callback: CodeRequestCallback?) {
        sendCode(authenticationCodeUrl, phoneNumber, callback)
    }

    override fun sendRegistrationCode(phoneNumber: String?, callback: CodeRequestCallback?) {
        sendCode(registrationCodeUrl, phoneNumber, callback)
    }

    override fun sendVerificationCode(phoneNumber: String?, callback: CodeRequestCallback?) {
        sendCode(verificationCodeUrl, phoneNumber, callback)
    }

    override fun verificationCode(
        phoneNumber: String?,
        code: String?,
        callback: VerificationCallback?
    ) {
        val myAccessToken = accessToken
        val myVerificationCodeUrl = verificationCodeUrl

        if (!myAccessToken.isPresent) {
            callback!!.onError(
                phoneNumber,
                KeycloakErrorResult("no token","must be have a token")
            )
            return
        }

        if(myVerificationCodeUrl == null){
            throw ExceptionInInitializerError("verificationCodeUrl not set!")
        }

        val url = HttpUrl.parse(myVerificationCodeUrl)?.newBuilder()?.
            addQueryParameter("phoneNumber", phoneNumber)?.
            addQueryParameter("code", code)?.
            build() ?: throw NullPointerException("verificationCodeUrl")

        val body = RequestBody.create(null, ByteArray(0))
        val request = Request.Builder()
            .addHeader("authorization", myAccessToken.orElseThrow { IllegalArgumentException() })
            .url(url)
            .post(body)
            .build()
        okHttpClient!!.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback!!.onFailure(phoneNumber)
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    callback!!.onSuccess(phoneNumber)
                } else {
                    callback!!.onFailure(phoneNumber)
                }
            }
        })
    }

    override fun requireToken(credential: Credential?, callback: AuthenticationCallback?) {
        val myClientId = clientId ?: throw ExceptionInInitializerError("client id not set!")
        val myAccessTokenUrl = accessTokenUrl ?: throw ExceptionInInitializerError("accessTokenUrl not set!")

        val formBody = FormBody.Builder().apply {
            add("client_id", myClientId)
            add("grant_type", "password")

            credential?.context?.forEach(BiConsumer { k: String?, v: String? -> add(k, v) })

            clientSecret?.let { add("client_secret", it) }
            scope?.let { add("scope", it) }
        }

        val request = Request.Builder()
            .addHeader("Content-Type", "application/x-www-form-urlencoded")
            .url(myAccessTokenUrl)
            .post(formBody.build())
            .build()
        okHttpClient?.newCall(request)?.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                Log.e(LOG_TAG, "failure", e)
                callback?.onFailure(KeycloakErrorResult(e.javaClass.simpleName, e.message))
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    assert(response.body() != null)
                    saveToken(parseToken(response.body()!!.string()))
                    callback?.onSuccess()
                } else {
                    if (response.body() != null) {
                        val body = response.body()?.string()
                        Log.e(LOG_TAG, body ?: "N/A")
                        callback?.onFailure(parseErrorResult(body))
                    } else {
                        Log.e(LOG_TAG, "unknown error")
                        callback?.onFailure(KeycloakErrorResult("unknown", "unknown error"))
                    }
                }
            }
        })
    }

    override fun refreshToken(): Optional<String> {
        return getToken().flatMap {
            refreshToken(token!!.refreshToken)
        }.map { t: KeycloakTokenInfo -> t.tokenType + " " + t.accessToken }
    }

    private fun refreshToken(refreshToken: String?): Optional<KeycloakTokenInfo> {
        Log.d("Keycloak", "refresh token")
        val myClientId = clientId ?: throw ExceptionInInitializerError("client id not set!")
        val myAccessTokenUrl = accessTokenUrl ?: throw ExceptionInInitializerError("accessTokenUrl not set!")

        Log.d("Keycloak", "refresh token for:$accessTokenUrl")
        val formBody = FormBody.Builder().apply {
            add("client_id", myClientId)
            add("grant_type", "refresh_token")
            add("refresh_token", refreshToken ?: "")

            clientSecret?.let { add("client_secret", it) }
            scope?.let { add("scope", it) }
        }

        val request = Request.Builder()
            .addHeader("Content-Type", "application/x-www-form-urlencoded")
            .url(myAccessTokenUrl)
            .post(formBody.build())
            .build()
        try {
            val response = okHttpClient!!.newCall(request).execute()
            if (response.isSuccessful) {
                Log.d(LOG_TAG, "token is refresh")
                assert(response.body() != null)
                val token = parseToken(response.body()!!.string())
                saveToken(token)
                return Optional.of(token)
            } else {
                if (response.body() != null) {
                    Log.e(
                        LOG_TAG, "refresh token fail! body is:" + response.body()!!
                            .string()
                    )
                } else {
                    Log.e(LOG_TAG, "refresh token fail! body is null!")
                }
                //remove this token? maybe only network fail!
            }
        } catch (e: IOException) {
            Log.e(LOG_TAG, "refresh token fail! body is:" + e.message, e)
        }
        return Optional.empty()
    }

    private fun sendCode(
        requestUrl: String?,
        phoneNumber: String?,
        callback: CodeRequestCallback?
    ) {
        val myRequestUrl = requestUrl ?: throw ExceptionInInitializerError("requestUrl not set!")
        val url = HttpUrl.parse(myRequestUrl)?.newBuilder()?.addQueryParameter("phoneNumber", phoneNumber)?.build()
        val request = Request.Builder()
            .addHeader("Accept", "application/json; charset=utf-8")
            .addHeader("Cache-Control", "no-cache")
            .url(url)
            .build()
        okHttpClient!!.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback?.onFailure(phoneNumber, KeycloakErrorResult("unknown"))
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    assert(response.body() != null)
                    callback?.onSuccess(phoneNumber, parseCodeResult(response.body()!!.string()))
                } else {
                    if (response.body() != null) {
                        callback?.onFailure(
                            phoneNumber,
                            parseErrorResult(response.body()!!.string())
                        )
                    } else {
                        callback?.onFailure(phoneNumber, KeycloakErrorResult("unknow"))
                    }
                }
            }
        })
    }

    private fun parseErrorResult(json: String?): KeycloakErrorResult {
        if (json == null || json.trim { it <= ' ' }.isEmpty()) {
            return KeycloakErrorResult("unknow")
        }
        val gsonBuilder = GsonBuilder()
        gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
        gsonBuilder.setLenient()
        return gsonBuilder.create().fromJson(json, KeycloakErrorResult::class.java)
    }

    private fun parseToken(json: String): KeycloakTokenInfo {
        Log.d("Keycloak", "parseToken$json")
        val gsonBuilder = GsonBuilder()
        gsonBuilder.setLenient()
        gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
        val result = gsonBuilder.create().fromJson(json, KeycloakTokenInfo::class.java)
        result.created = Date()
        return result
    }

    private fun parseCodeResult(json: String): KeycloakCodeResult {
        val gsonBuilder = GsonBuilder()
        gsonBuilder.setLenient()
        gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
        val result = gsonBuilder.create().fromJson(json, KeycloakCodeResult::class.java)
        result.created = Date()
        return result
    }

    companion object {
        private const val LOG_TAG = "KeycloakTokenHolder"
    }
}